File: app/shared/src/commonMain/kotlin/madang/shared/main/Git.kt
package madang.shared.main

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import madang.api.client.GitApi
import madang.api.model.GitBranches
import madang.api.model.GitCommitCreate
import madang.api.model.GitFile
import madang.api.model.GitLogEntry
import madang.api.model.GitStage
import madang.api.model.GitStatus
import madang.api.model.GitWorktree
import madang.shared.core.bodyOrThrow

/** git 탭 로그에 보이는 커밋 수. */
const val GIT_LOG_LIMIT = 20

/** 인덱스에 들어간 변경이 있다(`git status --porcelain`의 첫 글자). */
val GitFile.staged: Boolean get() = code.first() != ' ' && code != UNTRACKED

/** 작업 트리에 스테이징하지 않은 변경이 있다(둘째 글자, 추적 안 된 파일 포함). */
val GitFile.unstaged: Boolean get() = code.getOrElse(1) { ' ' } != ' ' || code == UNTRACKED

private const val UNTRACKED = "??"

/**
 * git 탭이 보이는 것. 저장소가 아니면 [status]만 있고 나머지는 비어 있다.
 *
 * @property branches 로컬 브랜치와 메인 체크아웃의 브랜치.
 * @property worktrees 메인 체크아웃과 워크트리, 그 워크트리를 쓰는 페이지.
 * @property log HEAD에서 거슬러 올라간 최근 커밋.
 */
data class GitView(
    val status: GitStatus,
    val branches: GitBranches? = null,
    val worktrees: List<GitWorktree> = emptyList(),
    val log: List<GitLogEntry> = emptyList()
) {
    val repository: Boolean get() = status.repository
}

/**
 * git 탭 상태. [project]가 null이면 닫혀 있다.
 *
 * @property page 열린 페이지. 있으면 그 페이지의 워크트리가 대상이다.
 * @property message 커밋 메시지 입력.
 * @property busy core에 보낸 조작이 끝나지 않았다.
 * @property error 마지막 조작을 core가 거부한 이유(정책 `denied`, 원격 없음 등).
 */
data class GitState(
    val project: String? = null,
    val page: String? = null,
    val view: Load<GitView>? = null,
    val message: String = "",
    val busy: Boolean = false,
    val error: String? = null
)

/**
 * git 탭. 상태·브랜치·워크트리·로그를 core에서 받고, 스테이지·커밋·푸시·풀·git 시작을 core에
 * 요청한다. 앱은 git을 직접 실행하지 않는다. 저장소가 아니면 상태만 받고 "git 시작"만 보인다.
 */
class GitViewModel(private val api: GitApi, private val scope: CoroutineScope) {

    private val _state = MutableStateFlow(GitState())
    val state: StateFlow<GitState> = _state.asStateFlow()

    fun open(project: String, page: String?) {
        val current = _state.value
        if (current.project == project && current.page == page) return
        _state.value = GitState(project = project, page = page)
        reload()
    }

    fun close() {
        _state.value = GitState()
    }

    fun setMessage(message: String) = _state.update { it.copy(message = message) }

    /** 다시 받는다. 받는 동안 앞의 내용을 그대로 보인다. */
    fun reload() {
        val (project, page) = target() ?: return
        _state.update { it.copy(view = it.view.takeIf { v -> v is Load.Ready } ?: Load.Loading) }
        scope.launch {
            val view = loadOf { fetch(project, page) }
            _state.update {
                if (it.project == project &&
                    it.page == page
                ) {
                    it.copy(view = view)
                } else {
                    it
                }
            }
        }
    }

    /** [paths]를 스테이징한다. null이면 모든 변경. */
    fun stage(paths: List<String>?) = act { project, page ->
        api.stageGit(project, GitStage(paths), page).bodyOrThrow()
    }

    /** 스테이징한 변경을 입력한 메시지로 커밋한다. 메시지가 비었으면 보내지 않는다. */
    fun commit() {
        val message = _state.value.message.trim()
        if (message.isEmpty()) return
        act { project, page ->
            api.commitGit(project, GitCommitCreate(message), page).bodyOrThrow()
            _state.update { if (it.message.trim() == message) it.copy(message = "") else it }
        }
    }

    fun push() = act { project, page -> api.pushGit(project, page).bodyOrThrow() }

    fun pull() = act { project, page -> api.pullGit(project, page).bodyOrThrow() }

    /** 프로젝트 폴더를 git 저장소로 만든다("git 시작"). */
    fun init() = act { project, _ -> api.initGit(project).bodyOrThrow() }

    /** `git.changed`. 보고 있는 프로젝트면 다시 받는다. */
    fun onChanged(project: String) {
        if (_state.value.project == project) reload()
    }

    private suspend fun fetch(project: String, page: String?): GitView {
        val status = api.getGitStatus(project, page).bodyOrThrow()
        if (!status.repository) return GitView(status)
        return GitView(
            status = status,
            branches = api.listGitBranches(project).bodyOrThrow(),
            worktrees = api.listGitWorktrees(project).bodyOrThrow(),
            log = api.getGitLog(project, page, limit = GIT_LOG_LIMIT).bodyOrThrow()
        )
    }

    /** core에 조작 하나를 보내고 끝나면 다시 받는다. 거부되면 이유를 보인다. */
    private fun act(block: suspend (project: String, page: String?) -> Unit) {
        val (project, page) = target() ?: return
        if (_state.value.busy) return
        _state.update { it.copy(busy = true, error = null) }
        scope.launch {
            val done = loadOf { block(project, page) }
            _state.update {
                if (it.project != project) return@update it
                it.copy(busy = false, error = (done as? Load.Failed)?.message)
            }
            reload()
        }
    }

    private fun target(): Pair<String, String?>? =
        _state.value.project?.let { it to _state.value.page }
}

File: core/madang/api/routes/git.py
"""git 경로: 앱 git 탭의 상태·디프·스테이지·커밋·브랜치·워크트리·이력·원격.

git은 core ``git`` 모듈만 실행한다. 모든 쓰기는 먼저 프로젝트 정책의
금지 명령(``policy.deny``)을 확인하고, 끝나면 ``git.changed``로 알린다.
``page``를 주면 그 페이지의 워크트리(있으면)에서 일한다.
"""

from __future__ import annotations

from pathlib import Path
from typing import Annotated

from fastapi import Query

from madang import git
from madang.api import errors, models, workspace
from madang.api.core import Core
from madang.api.routes import CoreDep, Router
from madang.cli_agent import repo as coderepo
from madang.cli_agent.context import AgentError
from madang.git import worktree
from madang.store import projects
from madang.store.page import worktree_path

# 계약의 태그. 목록 첫 자리에 두면 git 실행 검사에 걸리므로 이름으로 둔다.
TAG = "git"
router = Router(tags=[TAG])

PageQuery = Annotated[
    str | None,
    Query(description="이 페이지의 워크트리에서 일한다(있을 때)."),
]
LOG_LIMIT = 50


def _folder(
    core: Core, project: str, page: str | None
) -> tuple[projects.Project, Path]:
    found = core.project(project)
    page_dir = workspace.page_in(core, found, page)
    return found, workspace.work_folder(found, page_dir)


def _repo(
    core: Core, project: str, page: str | None
) -> tuple[projects.Project, Path]:
    found, folder = _folder(core, project, page)
    return found, workspace.require_repo(folder)


def _git_failed(exc: git.GitError) -> errors.HttpError:
    return errors.conflict(str(exc))


def git_status(folder: Path) -> models.GitStatus:
    """작업 트리의 상태. git 저장소가 아니면 ``repository``가 거짓이다."""
    if not git.is_work_tree(folder):
        return models.GitStatus(repository=False, folder=str(folder), files=[])
    try:
        head = git.rev_parse(folder)
    except git.GitError:
        head = None  # 아직 커밋이 없다
    files = [
        models.GitFile(path=path, code=code)
        for path, code in sorted(git.status(folder).items())
    ]
    return models.GitStatus(
        repository=True,
        folder=str(folder),
        branch=git.current_branch(folder),
        head=head,
        files=files,
    )


@router.get("/projects/{project}/git/status", operation_id="getGitStatus")
def get_status(
    project: str, core: CoreDep, page: PageQuery = None
) -> models.GitStatus:
    """변경된 파일, 브랜치, HEAD."""
    _, folder = _folder(core, project, page)
    try:
        return git_status(folder)
    except git.GitError as exc:
        raise _git_failed(exc) from exc


@router.get("/projects/{project}/git/diff", operation_id="getGitDiff")
def get_diff(
    project: str,
    core: CoreDep,
    page: PageQuery = None,
    staged: bool = False,
    base: str | None = None,
    target: str | None = None,
    path: Annotated[list[str] | None, Query()] = None,
) -> models.GitDiff:
    """통합 diff 텍스트."""
    _, repo = _repo(core, project, page)
    try:
        text = git.diff(repo, base, target, staged=staged, paths=path or ())
    except git.GitError as exc:
        raise _git_failed(exc) from exc
    return models.GitDiff(text=text)


@router.post("/projects/{project}/git/stage", operation_id="stageGit")
def stage(
    project: str,
    body: models.GitStage,
    core: CoreDep,
    page: PageQuery = None,
) -> models.GitStatus:
    """변경을 스테이징한다. ``paths``가 없으면 모든 변경이다."""
    found, repo = _repo(core, project, page)
    workspace.allow(found, "git add -A")
    try:
        git.stage(repo, body.paths)
        status = git_status(repo)
    except git.GitError as exc:
        raise _git_failed(exc) from exc
    core.announce_git(found.id, repo, "stage")
    return status


@router.post("/projects/{project}/git/commit", operation_id="commitGit")
def commit(
    project: str,
    body: models.GitCommitCreate,
    core: CoreDep,
    page: PageQuery = None,
) -> models.RepoCommitResult:
    """스테이징된 변경을 커밋한다. ``paths``를 주면 그 경로만 커밋한다."""
    found, repo = _repo(core, project, page)
    if not body.message.strip():
        raise errors.invalid("commit message is empty")
    workspace.allow(found, "git commit")
    blocked = coderepo.sensitive(list(git.status(repo)))
    if blocked:
        raise errors.conflict(
            "files that may hold secrets are not committed: "
            + ", ".join(blocked)
        )
    if not body.paths and not git.has_staged_changes(repo):
        raise errors.conflict("nothing is staged")
    try:
        sha = git.commit(repo, body.message, body.paths)
    except git.GitError as exc:
        raise _git_failed(exc) from exc
    core.announce_git(found.id, repo, "commit", commit=sha)
    return models.RepoCommitResult(commit=sha)


@router.get("/projects/{project}/git/branches", operation_id="listGitBranches")
def list_branches(project: str, core: CoreDep) -> models.GitBranches:
    """로컬 브랜치와 메인 체크아웃의 현재 브랜치."""
    _, repo = _repo(core, project, None)
    try:
        return models.GitBranches(
            current=git.current_branch(repo), branches=git.branches(repo)
        )
    except git.GitError as exc:
        raise _git_failed(exc) from exc


@router.get(
    "/projects/{project}/git/worktrees", operation_id="listGitWorktrees"
)
def list_worktrees(project: str, core: CoreDep) -> list[models.GitWorktree]:
    """메인 체크아웃과 페이지 워크트리. 페이지 워크트리에는 페이지 id가 있다."""
    found, repo = _repo(core, project, None)
    try:
        trees = worktree.list_worktrees(repo)
    except git.GitError as exc:
        raise _git_failed(exc) from exc
    pages = {
        worktree_path(found.root, p.name).resolve(): p.name
        for p in (found.pages_dir.iterdir() if found.pages_dir.is_dir() else ())
    }
    return [
        models.GitWorktree(
            path=str(tree.path),
            branch=tree.branch,
            head=tree.head,
            main=tree.path.resolve() == found.root.resolve(),
            page=pages.get(tree.path.resolve()),
        )
        for tree in trees
    ]


@router.get("/projects/{project}/git/log", operation_id="getGitLog")
def get_log(
    project: str,
    core: CoreDep,
    page: PageQuery = None,
    ref: str = "HEAD",
    limit: Annotated[int, Query(ge=1, le=500)] = LOG_LIMIT,
) -> list[models.GitLogEntry]:
    """``ref``에서 거슬러 올라가는 커밋, 최근 것부터."""
    _, repo = _repo(core, project, page)
    try:
        commits = git.log(repo, ref, limit)
    except git.GitError as exc:
        raise errors.not_found(str(exc)) from exc
    return [
        models.GitLogEntry(
            hash=c.hash, author=c.author, date=c.date, subject=c.subject
        )
        for c in commits
    ]


@router.post("/projects/{project}/git/push", operation_id="pushGit")
def push(
    project: str, core: CoreDep, page: PageQuery = None
) -> models.RepoPushResult:
    """현재 브랜치를 같은 이름의 원격 브랜치로 강제 없이 보낸다."""
    found, repo = _repo(core, project, page)
    return push_folder(core, found, repo)


def push_folder(
    core: Core, project: projects.Project, repo: Path
) -> models.RepoPushResult:
    """정책을 확인하고 ``repo``의 현재 브랜치를 push한다.

    Raises:
        HttpError: 정책이 막았거나(409 ``denied``) push할 수 없다(409).
    """
    branch = git.current_branch(repo) or "HEAD"
    workspace.allow(project, f"git push origin {branch}")
    try:
        remote, branch = coderepo.push_current(repo)
    except AgentError as exc:
        raise errors.conflict(str(exc)) from exc
    core.announce_git(project.id, repo, "push", branch=branch)
    return models.RepoPushResult(branch=branch, remote=remote)


@router.post("/projects/{project}/git/pull", operation_id="pullGit")
def pull(
    project: str, core: CoreDep, page: PageQuery = None
) -> models.GitStatus:
    """빨리 감기로만 당겨 온다. 갈라졌으면 409."""
    found, repo = _repo(core, project, page)
    workspace.allow(found, "git pull --ff-only")
    try:
        git.pull(repo)
        status = git_status(repo)
    except git.GitError as exc:
        raise _git_failed(exc) from exc
    core.announce_git(found.id, repo, "pull", head=status.head)
    return status


@router.post(
    "/projects/{project}/git/init", status_code=201, operation_id="initGit"
)
def init(project: str, core: CoreDep) -> models.GitStatus:
    """프로젝트 폴더를 git 저장소로 만든다. ``.madang/``은 커밋에서 뺀다."""
    found, folder = _folder(core, project, None)
    if git.is_work_tree(folder):
        raise errors.conflict(f"{folder} is already a git repository")
    workspace.allow(found, "git init")
    try:
        git.run(folder, "init", "-q")
        # 기록 폴더를 track 설정대로 .git/info/exclude에 더한다.
        projects.create_records(found.root)
        status = git_status(folder)
    except (git.GitError, OSError, ValueError) as exc:
        raise errors.conflict(f"cannot initialize git: {exc}") from exc
    core.announce_git(found.id, folder, "init")
    return status
